### PR TITLE
feature/ST-348-bypass-cardinal-if-non-3d-card-type

### PR DESCRIPTION
**Pull request by pgs-gmartowski**

### CODE REVIEW

File: src/core/services/Config.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    'errorcode',
+    'errordata',
+    'errormessage',
+    'orderreference',
+    'settlestatus',
+    'status',
+    'transactionreference'
+  ];
+
+  public init(config: IConfig) {
```
Comment (wirus15) [inline src/core/services/Config.ts]: return types are missing

File: src/core/services/Config.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+      defaultPaymentType:
+        config.components.defaultPaymentType !== undefined ? config.components.defaultPaymentType : '',
+      paymentTypes: config.components.paymentTypes !== undefined ? config.components.paymentTypes : [''],
+      requestTypes:
+        config.components.requestTypes !== undefined ? config.components.requestTypes : ['THREEDQUERY', 'AUTH'],
+      startOnLoad: config.components.startOnLoad !== undefined ? config.components.startOnLoad : false
+    };
+  }
+
+  private _setApmConfig(apm: IWalletConfig, components: IComponentsConfig): IWalletConfig {
```
Comment (wirus15) [inline src/core/services/Config.ts]: if `apm` can be empty it should be noted in argument (and also return type here). 
```typescript
private _setApmConfig(?apm: IWalletConfig, components: IComponentsConfig): ?IWalletConfig {
    if (!apm) {
        return apm;
    }

    return { ... };
}

Comment (pgs-gmartowski) [reply src/core/services/Config.ts]: done

File: src/core/services/Config.ts (hunk 1 above)
Comment (pgs-gmartowski) [reply src/core/services/Config.ts]: changed

File: src/core/models/Config.ts [hunk 3]
```diff
@@ -109,4 +143,4 @@ const IComponentsConfigSchema = Joi.object().keys({
   startOnLoad: Joi.boolean()
 });
 
-export { IByPassInit, IConfig, IConfigSchema, IComponentsConfig, IComponentsConfigSchema, IWalletConfig };
+export { IByPassInit, IConfig, IComponentsConfig, IComponentsConfigSchema, IWalletConfig, ByPassCards, IComponentsIds };
```
[the remark this answers is not in the document]
Comment (pgs-gmartowski) [reply src/core/models/Config.ts]: I think this should be resolved after merging other branches with refactors

File: src/core/models/Config.ts [hunk 4]
```diff
@@ -1,11 +1,34 @@
 import Joi from 'joi';
 import { IStyles } from '../shared/Styler';
 
+enum ByPassCards {
```
[the remark this answers is not in the document]
Comment (pgs-gmartowski) [reply src/core/models/Config.ts]: should be fine now

File: src/ST.ts [hunk 5]
```diff
@@ -15,337 +14,134 @@ import CardinalCommerceMock from './core/integrations/CardinalCommerceMock';
...
-      origin,
-      componentIds,
-      styles,
-      submitOnSuccess,
-      submitOnError,
-      submitFields,
-      gatewayUrl,
-      animatedCard,
-      submitCallback
+  private Init(config: IConfig): void {
```
[the remark this answers is not in the document]
Comment (pgs-gmartowski) [reply src/ST.ts]: ok, fixed

File: src/ST.ts [hunk 6]
```diff
@@ -15,337 +14,134 @@ import CardinalCommerceMock from './core/integrations/CardinalCommerceMock';
...
-          'errormessage',
-          'orderreference',
-          'settlestatus',
-          'status',
-          'transactionreference'
-        ];
-    return defaultSubmitFields;
+  public ApplePay(): ApplePay {
+    const { applepay } = this.Environment();
+    return new applepay(this._config.applePay, this._config.jwt, this._config.datacenterurl);
```
[the remark this answers is not in the document]
Comment (pgs-gmartowski) [reply src/ST.ts]: fixed

File: example/index.html [hunk 7]
```diff
@@ -138,11 +138,11 @@ <h1 class="example-form__title">
...
           };
-          st.Components(stConfig.components);
-          st.ApplePay(stConfig.applePay);
-          st.VisaCheckout(stConfig.visaCheckout);
+          st.Components();
+          st.ApplePay();
+          st.VisaCheckout();
           document.getElementById('example-form-amount').addEventListener('input', function() {
-            st.updateJWT(
+            st.UpdateJwt(
```
Comment (james-simpson-ST) [inline example/index.html]: As we have already published updateJWT instead of UpdateJwt we can't change this interface for the merchant unfortunately so we need to revert this change

Comment (pgs-gmartowski) [reply example/index.html]: Yes, I'm sorry I forgot about this :/
Reverted.

File: src/core/services/Storage.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    }
+  }
+
+  /**
+   * Sets data to localStorage, previously converts
+   * them to string depends on type given.
+   * @param name
+   * @param storage
+   */
+  public getLocalStorageItem(name: string, storage: string) {
```
[the remark this answers is not in the document]
Comment (pgs-gmartowski) [reply src/core/services/Storage.ts]: done

File: src/core/services/Storage.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
+export class Storage {
```
[the remark this answers is not in the document]
Comment (pgs-gmartowski) [reply src/core/services/Storage.ts]: changed

File: src/core/services/Storage.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  /**
+   * Sets data to localStorage, previously converts
+   * them to string depends on type given.getPaymentSuccessStatus
+   * @param name
+   * @param data
+   */
+  public setLocalStorageItem(name: string, data: any) {
```
[the remark this answers is not in the document]
Comment (pgs-gmartowski) [reply src/core/services/Storage.ts]: ok, done

File: example/index.html [hunk 11]
```diff
@@ -138,9 +138,9 @@ <h1 class="example-form__title">
             var stringified = JSON.stringify(data);
             var testVariable = 'This is what we have got after submit' + stringified;
           };
-          st.Components(stConfig.components);
-          st.ApplePay(stConfig.applePay);
-          st.VisaCheckout(stConfig.visaCheckout);
+          st.Components();
```
Comment (james-simpson-ST) [inline example/index.html]: Doesn't example page still need to pass in stConfig for Components,  ApplePay and VisaCheckout?

Comment (pgs-gmartowski) [reply example/index.html]: nope, merchant should now provide whole config object to 
`var st = ST(stConfig)`;
I can go back to previous configuration if you want; I thought my solution would be better but maybe I went too far ;)

Comment (james-simpson-ST) [reply example/index.html]: Unfortunately, as merchants are already passing in the config to Components and ApplePay we can't change this (again this would be a backwards-incompatible change and we can't have them :(

Comment (pgs-gmartowski) [reply example/index.html]: :(
Ok, so I'm restoring previous configuration then

Comment (pgs-gmartowski) [reply example/index.html]: Done, `index.html` should be ok now. 
@msakowicz could you check it once again ?
